Replace `PCtx` parameters with `TCtx` in VM code

zerbina opened the pull request:
* Where `TCtx` is now used, also use `func` instead of `proc`, if applicable

Not all usages of `PCtx` are replaced. `PCtx` is still used in procedures
that form the border between non-VM/VM code. `vmops` also still uses
 `PCtx`, as some ops need to capture a reference to the context

zerbina committed:
Replace `PCtx` parameters with `TCtx` in VM code

* Where `TCtx` is now used, also use `func` instead of `proc`, if applicable

Not all usages of `PCtx` are replaced. `PCtx` is still used in procedures
that form the border between non-VM/VM code. `vmops` also still uses
 `PCtx`, as some ops need to capture a reference to the context

alaviss commented on compiler/vm/vm.nim:
```diff
@@ -540,7 +540,7 @@ proc recSetFlagIsRef(arg: PNode) =
     arg[i].recSetFlagIsRef
 
-proc setLenSeq(c: PCtx; node: PNode; newLen: int; info: TLineInfo) =
+func setLenSeq(c: TCtx; node: PNode; newLen: int; info: TLineInfo) =
```
The compiler has spoken :P

```suggestion
proc setLenSeq(c: TCtx; node: PNode; newLen: int; info: TLineInfo) =
```

saem commented on compiler/vm/vmdef.nim:
```diff
@@ -189,17 +189,22 @@ proc newCtx*(module: PSym; cache: IdentCache; g: ModuleGraph; idgen: IdGenerator
   result = c.callbacks.len
   c.callbacks.add((name, callback))
 
+# For a transition period, keep an overload of `registerCallback` that still
```
Maybe mark the template as deprecated with the current comment as the deprecated warning? Either that or make it a doc comment instead of a comment.

saem reviewed (changes requested):
Minor item, the other thing I noticed is that some functions have effects, will need to change those to procs or break them up. I'd go the proc route, personally.

zerbina commented:
Addressed review. I used `proc` again for the functions that fail the strict side-effect test.

I also fixed the failure of the ``tcompilerapi.nim`` test. It failed due to `PCtx` still being used as arguments in ``nimeval.nim``.

Interestingly enough, in the CI, the test failure was reported to be due to the compiler crashing with a nil access. I ran the test on my machine (Windows) and it also failed with a compiler crash. But, instead of an invalid memory access, I got an assertion raised in ``semcall.nim``.

 The compiler crash only happened when importing `nimeval`. Adjusting the `PCtx` usage made the crash go away too.

saem commented:
I did a quick search in finding usage in semdata and semmagic, but those are at the boundary as mentioned by the PR.

saem reviewed (approved):
And so begins the revamp of the VM in devel. 🎉 

Bors r+

alaviss commented:
Weird, let's try again

bors r+
